rfcs/XXXX-gating/README.md

```diff
@@ -0,0 +1,301 @@
+# RFC-XXXX Gating Flux reconciliation
+
+**Status:** provisional
+
+**Creation date:** 2022-09-28
+
+**Last update:** 2022-10-04
+
+## Summary
+
+Flux should offer a mechanism for cluster admins and other teams involved in the release process
+to manually approve the rollout of changes onto clusters. In addition, Flux should offer 
+a way to define maintenance time windows and other time-based gates, to allow a better control 
+of applications and infrastructure changes to critical system.
+
+## Motivation
+
+Flux watches sources (e.g. GitRepositories, OCIRepositories, HelmRepositories, S3-compatible Buckets) and
+automatically reconciles the changes onto clusters as described with Flux Kustomizations and HelmReleases.
+The teams involved in the delivery process (e.g. dev, qa, sre) can decide when changes are delivered
+to production by reviewing and approving the proposed changes in a collaborative manner with pull request.
+Once a pull request is merged onto a branch that defines the desired state of the production system,
+Flux kicks off the reconciliation process.
+
+There are situations when users want to have a gating mechanism after the desired state changes are merged in Git:
+
+- Manual approval of container image updates (e.g. https://github.com/fluxcd/flux2/discussions/870)
+- Manual approval of infrastructure upgrades (e.g. https://github.com/fluxcd/flux2/issues/959)
+- Maintenance window (e.g. https://github.com/fluxcd/flux2/discussions/1004)
+- Planned releases
+- No Deploy Friday
+
+### Goals
+
+- Offer a dedicated API for defining time-based gates in a declarative manner.
+- Introduce a `gating-controller` in the Flux suite that manages the `Gate` objects.
+- Extend the current Flux APIs and controllers to support gating.
+
+### Non-Goals
+
+<!--
+What is out of scope for this RFC? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+In order to support manual gating, Flux could be extended with a dedicated API and controller
+that would allow users to define `Gate` objects and perform operations like `open` and `close`.
+
+A `Gate` object could be referenced in sources (Buckets, Git, Helm, OCI Repositories)
+and syncs (Kustomizations, HelmReleases, ImageUpdateAutomation)
+to block the reconciliation until the gate is opened.
+
+A `Gate` can be opened or closed by annotating the object with a timestamp or by
+calling a specific webhook receiver exposed by notification-controller.
+
+A `Gate` can be configured to automatically close or open based on a time window defined in the `Gate` spec.
+
+The `Gate` API would replace Flagger's current
+[manual gating mechanism](https://docs.flagger.app/usage/webhooks#manual-gating).
+
```

Comment:
Additional audit trail for this feature:
- Enriching existing logs so that actions that can be gated should log:
	- What gate(s) were checked (if any).
	- What gate(s) caused the final result of (not) going ahead with the action.
	- If multiple gates can be defined, changes to which gates are being observed and what logical operators (and/or) are being used, should be descriptively logged.
- Logs for the new Gate controller:
	- Descriptive summary of the gate state at creation time.
	- Any change of Gate state should log a descriptive summary.
- The additional logs above would be an "opt-in" feature, so that users that do not require this level of governance don't have to process the extra data.